src/topology/continuous_function/stone_weierstrass.lean

```diff
@@ -0,0 +1,353 @@
+begin
+  rw inf_eq,
+  refine A.topological_closure.smul_mem
+    (A.topological_closure.sub_mem
+      (A.topological_closure.add_mem (A.subalgebra_topological_closure f.property)
+          (A.subalgebra_topological_closure g.property)) _) _,
+  exact_mod_cast abs_mem_subalgebra_closure A _,
+end
+
+theorem inf_mem_closed_subalgebra (A : subalgebra ℝ C(X, ℝ)) (h : is_closed (A : set C(X, ℝ)))
+  (f g : A) : (f : C(X, ℝ)) ⊓ (g : C(X, ℝ)) ∈ A :=
+begin
+  convert inf_mem_subalgebra_closure A f g,
+  apply set_like.ext',
+  symmetry,
+  erw closure_eq_iff_is_closed,
+  exact h,
+end
+
+theorem sup_mem_subalgebra_closure (A : subalgebra ℝ C(X, ℝ)) (f g : A) :
+  (f : C(X, ℝ)) ⊔ (g : C(X, ℝ)) ∈ A.topological_closure :=
+begin
+  rw sup_eq,
+  refine A.topological_closure.smul_mem
+    (A.topological_closure.add_mem
+      (A.topological_closure.add_mem (A.subalgebra_topological_closure f.property)
+          (A.subalgebra_topological_closure g.property)) _) _,
+  exact_mod_cast abs_mem_subalgebra_closure A _,
+end
+
+theorem sup_mem_closed_subalgebra (A : subalgebra ℝ C(X, ℝ)) (h : is_closed (A : set C(X, ℝ)))
+  (f g : A) : (f : C(X, ℝ)) ⊔ (g : C(X, ℝ)) ∈ A :=
+begin
+  convert sup_mem_subalgebra_closure A f g,
+  apply set_like.ext',
+  symmetry,
+  erw closure_eq_iff_is_closed,
+  exact h,
+end
+
+open_locale topological_space
+
+-- Here's the fun part of Stone-Weierstrass!
+theorem sublattice_closure_eq_top
+  (L : set C(X, ℝ)) (nA : L.nonempty)
+  (inf_mem : ∀ f g ∈ L, f ⊓ g ∈ L) (sup_mem : ∀ f g ∈ L, f ⊔ g ∈ L)
+  (sep : L.separates_points_strongly) :
+  closure L = ⊤ :=
+begin
+  -- We start by boiling down to a statement about close approximation.
+  apply eq_top_iff.mpr,
+  rintros f -,
+  refine filter.frequently.mem_closure
+    ((filter.has_basis.frequently_iff metric.nhds_basis_ball).mpr (λ ε pos, _)),
+  simp only [exists_prop, metric.mem_ball],
+
+  -- It will be helpful to assume `X` is nonempty later,
+  -- so we get that out of the way here.
+  by_cases nX : nonempty X,
+  swap,
+  exact ⟨nA.some, (dist_lt_iff _ _ pos).mpr (λ x, false.elim (nX ⟨x⟩)), nA.some_spec⟩,
+
+  /-
+  The strategy now is to pick a family of continuous functions `g x y` in `A`
+  with the property that `g x y x = f x` and `g x y y = f y`
+  (this is immediate from `h : separates_points_strongly`)
+  then use continuity to see that `g x y` is close to `f` near both `x` and `y`,
+  and finally using compactness to produce the desired function `h`
+  as a maximum over finitely many `x` of a minimum over finitely many `y` of the `g x y`.
+  -/
+  dsimp [set.separates_points_strongly] at sep,
+
+  let g : Π x y, L := λ x y, (sep f x y).some,
+  let w₁ : ∀ x y, g x y x = f x := λ x y, (sep f x y).some_spec.1,
+  let w₂ : ∀ x y, g x y y = f y := λ x y, (sep f x y).some_spec.2,
+
+  -- For each `x y`, we define `U x y` to be `{ z | f z - ε < g x y z }`,
+  -- and observe this is a neighbourhood of `y`.
+  let U : Π x y, set X := λ x y, { z | f z - ε < g x y z },
```

Comment:
```suggestion
  let U : Π x y, set X := λ x y, {z | f z - ε < g x y z},
```
I think we don't put spaces around braces when defining sets. Same thing in the docstring two lines above, and later in the proof.